contributors/design-proposals/grow-volume-size.md

```diff
@@ -0,0 +1,253 @@
+
+    if err != nil {
+        return nil, fmt.Errorf("Error creating expander plugin for volume %q with error %v", pvcWithResizeRequest.UniquePvcKey(), err)
+    }
+
+    expandFunc := func() error {
+        expandErr := expanderPlugin.ExpandVolumeDevice(pvcWithResizeRequest.VolumeSpec, pvcWithResizeRequest.ExpectedSize, pvcWithResizeRequest.CurrentSize)
+
+        if expandErr != nil {
+            glog.Errorf("Error expanding volume through cloudprovider : %v", expandErr)
+            return expandErr
+        }
+        dsow.MarkAsResized(pvcWithResizeRequest)
+
+        return nil
+    }
+    return expandFunc, nil
+}
+```
+
+* Once volume expand is successful, the volume will be marked as expanded and new size will be updated in `pv.spec.capacity`. Any errors will be reported as *events* on PVC object.
+* If resize failed in above step, in addition to events - `pvc.Status.Conditions` will be updated with `ResizeFailed: True`. Corresponding error will be added to condition field as well.
+* Depending on volume type next steps would be:
+
+    * If volume is of type that does not require file system resize, then `pvc.status.capacity` will be immediately updated to reflect new size. This would conclude the volume expand operation. Also `pvc.Status.Conditions` will be updated with `Ready: True`.
+    * If volume if of type that requires file system resize then a file system resize will be performed on kubelet. Read below for steps that will be performed for file system resize.
+
+* If volume plugin is of type that can not do resizing of attached volumes (such as `Cinder`) then `ExpandVolumeDevice` can return error by checking for
+  volume status with its own API (such as by making Openstack Cinder API call in this case). Controller will keep trying to resize the volume until it is
+  successful.
+
+* To consider cases of missed PVC update events, an additional loop will reconcile bound PVCs with PVs. This additional loop will loop through all PVCs
+  and match `pvc.spec.resources.requests` with `pv.spec.capacity` and add PVC in `volume_expand_controller`'s desired state of world if `pv.spec.capacity` is less
+  than `pvc.spec.resources.requests`.
+
+* There will be additional checks in controller that grows PV size - to ensure that we do not make volume plugin API calls that can reduce size of PV.
+
+### File system resize on kublet
+
+A File system resize will be pending on PVC until a new pod that uses this volume is scheduled somewhere. While theoretically we *can* perform
+online file system resize if volume type and file system supports it - we are leaving it for next iteration of this feature.
+
+* When calling `MountDevice` or `Setup` call of volume plugin, volume manager will in addition compare `pv.spec.capacity` and `pvc.status.capacity` and if `pv.spec.capacity` is greater
+  than `pvc.status.spec.capacity` then volume manager will additionally resize the file system of volume.
```

Comment:
It is kubelet volume manager that is doing fs resize, instead of volume plugin?  How does it know what type of fs the volume has, and which tools to use?